price-discovery/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+        } else if payment_token == launched_token_id {
+            LAUNCHED_TOKEN_REDEEM_NONCE
+        } else {
+            return sc_error!(INVALID_PAYMENT_ERR_MSG);
+        };
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, redeem_token_nonce, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn withdraw(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(payment_token == redeem_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let refund_token_id = match payment_nonce {
+            LAUNCHED_TOKEN_REDEEM_NONCE => self.launched_token_id().get(),
+            ACCEPTED_TOKEN_REDEEM_NONCE => self.accepted_token_id().get(),
+            _ => return sc_error!(INVALID_PAYMENT_ERR_MSG),
+        };
+
+        self.burn_redeem_token(payment_nonce, &payment_amount);
+
+        let caller = self.blockchain().get_caller();
+        self.send()
+            .direct(&caller, &refund_token_id, 0, &payment_amount, &[]);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn redeem(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_deposit_period_ended()?;
+        require!(!self.lp_token_id().is_empty(), "Pool not created yet");
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(payment_token == redeem_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        self.burn_redeem_token(payment_nonce, &payment_amount);
+
+        let lp_token_amount = self.compute_lp_amount_to_send(payment_nonce, payment_amount);
+        require!(lp_token_amount > 0u32, "Nothing to redeem");
+
+        let caller = self.blockchain().get_caller();
+        let lp_token_id = self.lp_token_id().get();
+        self.send()
+            .direct(&caller, &lp_token_id, 0, &lp_token_amount, &[]);
+
+        Ok(())
+    }
+
+    // private
+
+    fn require_active(&self) -> SCResult<()> {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let start_epoch = self.start_epoch().get();
+        let end_epoch = self.end_epoch().get();
+        require!(
+            start_epoch <= current_epoch,
+            "Deposit period not started yet"
+        );
+        require!(current_epoch < end_epoch, "Deposit period ended");
+
+        Ok(())
+    }
+
+    fn compute_lp_amount_to_send(
+        &self,
+        redeem_token_nonce: u64,
+        redeem_token_amount: BigUint,
+    ) -> BigUint {
+        let total_lp_tokens = self.total_lp_tokens_received().get();
+
+        match redeem_token_nonce {
```

Comment:
we need to make a common code with swap here - as this is critical to be the same. Right now it looks ok.

Reply:
Done. Now it's using the `quote` function from the `amm` module: https://github.com/ElrondNetwork/sc-dex-rs/blob/main/dex/pair/src/amm.rs#L16

Reply:
Changed to use `rule_of_three` instead of `quote`